### PR TITLE
Add the ability to reorder array elements from the inspector

### LINKED ISSUE #23: Ability to change order of elements in export (Array)
**Describe the project you are working on:**

Any project that uses an export (Array). for example an export (Array, PackedScene)

**Describe how this feature / enhancement will help your project:**

If I need to iterate through array, but need it to be in a certain order, and I want to add an element (in this case a packedscene) at a certain index. Right now I have to erase every element until I get to the index I want, insert the new element, then reinsert every element after.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
Current:
export_array

There are a few ways to add this, for example:
-arrows next to the element index number to move it up or down the array
-Being able to drag the elements to sort them (however would not work with more than 1 page)
-Editable index number that will re-arrange elements according to the number you change it to
-"Insert at index #" option at top of the array
export_array_new

**Describe implementation detail for your proposal (in code), if possible:**

Sorry, I'm not sure how it would be implemented in code :(

**If this enhancement will not be used often, can it be worked around with a few lines of script?:** 

Possibly, for example by making a new export var and, in code, inserting that var to the export array at the index you want, but the point of exports are to be able to change it through the inspector right?

**Is there a reason why this should be core and not an add-on in the asset library?:**

Yes, this is a enhancement to the already implemented export arrays, which makes them easier to edit.

### PR DESCRIPTION
**Pull request by AaronRecord**
### Demo video:

https://user-images.githubusercontent.com/50304111/113367098-6355d600-9318-11eb-9a28-4360bd619f24.mp4

You click and drag to move elements up or down
Supports dragging between different pages.

Resolves https://github.com/godotengine/godot-proposals/issues/23,
Tested on 3.x (laptop doesn't support vulkan so I can't personally test on 4.0)

If it'd be helpful, I'm willing to open a separate PR for 3.x

### CODE REVIEW

**Comment by fire** [comment]
This is a great idea. Need to test it when I have time.

File: editor/editor_properties_array_dict.cpp [hunk 1]
```diff
@@ -384,12 +405,7 @@ void EditorPropertyArray::_remove_pressed(int p_index) {
 	Variant array = object->get_array();
 	array.call("remove", p_index);
 
-	if (array.get_type() == Variant::ARRAY) {
-		array = array.call("duplicate");
```
Comment (LightningAA) [inline editor/editor_properties_array_dict.cpp:388]: There are duplicate calls all over and I don't understand why. Some of them have a comment saying it's so undo/redo work better, but I tried commenting it out here and undo/redo still works fine for removing items.

File: editor/editor_properties_array_dict.cpp [hunk 2]
```diff
@@ -566,6 +587,70 @@ void EditorPropertyArray::setup(Variant::Type p_array_type, const String &p_hint
...
+			EditorNode::get_singleton()->get_inspector()->ensure_control_visible(reorder_selected_property_hbox);
+		}
+	}
+}
+
+void EditorPropertyArray::_reorder_button_down(int p_index) {
+	reorder_move_from_idx = p_index;
+	reorder_move_to_idx = p_index;
+	reorder_selected_property_hbox = Object::cast_to<HBoxContainer>(vbox->get_child(p_index % page_len + 2));
+	reorder_selected_button = Object::cast_to<Button>(reorder_selected_property_hbox->get_child(0));
```
Comment (LightningAA) [inline editor/editor_properties_array_dict.cpp:624]: My original plan was to use [mouse_exited](https://docs.godotengine.org/en/stable/classes/class_control.html#signals) on the property hbox, but it was called at the wrong time (https://github.com/godotengine/godot/issues/16854) or not at all.My original plan was to use [mouse_exited](https://docs.godotengine.org/en/stable/classes/class_control.html#signals) on the property hbox, but it was called at the wrong time (https://github.com/godotengine/godot/issues/16854) or not at all.

**Comment by LightningAA** [comment]
@fire Have you gotten around to test it yet? 😃

**Review by fire — APPROVED** [review]
Design looks good.

I used the built editor and it seems to work on an array.

Did a timelimited review of the code.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by LightningAA** [comment]
This should be able to be pretty easily cherry-picked, I don't think it'll require a separate PR.

**Comment by akien-mga** [comment]
> This should be able to be pretty easily cherry-picked, I don't think it'll require a separate PR.

There's a fair amount of conflicts, a PR would be useful.

**Comment by LightningAA** [comment]
@akien-mga Sorry about that, I figured it'd be fine since I originally developed it on the 3.x branch and forward ported it to 4.0, but I've been meaning to double check.

**Pull request merged**